File: lib/category-label.ts
/** Lowercase compare for Russian category labels. */
export function normalizeCategoryLabel(name: string): string {
  return name.toLowerCase().replace(/\s+/g, " ").trim()
}

/** CashPack title case → readable Russian (sentence case, «и» lowercase). */
export function formatCategoryLabel(name: string): string {
  return name
    .split(/\s+/)
    .map((word, index) => {
      if (word.toLowerCase() === "и" && index > 0) return "и"
      if (/^\(.+\)$/.test(word)) return word
      const lower = word.toLowerCase()
      return lower.charAt(0).toUpperCase() + lower.slice(1)
    })
    .join(" ")
}

export function labelsEquivalent(a: string, b: string): boolean {
  return normalizeCategoryLabel(a) === normalizeCategoryLabel(b)
}

File: backend/services/category_label.py
"""Mirror of lib/category-label.ts for matrix merge on the server."""

import re


def normalize_category_label(name: str) -> str:
    return re.sub(r"\s+", " ", name.lower()).strip()


def format_category_label(name: str) -> str:
    words = name.split()
    result: list[str] = []
    for index, word in enumerate(words):
        if word.lower() == "и" and index > 0:
            result.append("и")
            continue
        if re.fullmatch(r"\(.+\)", word):
            result.append(word)
            continue
        lower = word.lower()
        result.append(lower[0].upper() + lower[1:] if lower else word)
    return " ".join(result)


def labels_equivalent(a: str, b: str) -> bool:
    return normalize_category_label(a) == normalize_category_label(b)
